
# Exploring eBay Car Sales Data

In this project I'll be cleaning, preparing, and researching used car sales data. The data is sourced from a German eBay site (*eBay Kleinanzeigan*).

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')

Looks like we have a count of 50,000 rows and 20 columns to work with. I'm probably going to definitely make sue the `dateCrawled` index is in a consistent format. `name` will need to be broken down into make, model, and manufacturer (which looks synonymous with `brand`).

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

Columns `price` and `odometer` will need to be converted to `int64` or `float64` as needed. It will also be nice to change all our column headers to a more consistently pythonic version. There is also quite a bit of missing info in certain columns (particularly `notRepairedDamage`).

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

**Did Pandas add an index column or do I need to do that?**

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Columns are `camelCase` and not our lovely, Pythonic `snakecase`. We'll also convert oddly named columns like `nrOfPictures` to `picture_count` and the ambiguous `dateCreated` to `ad_created`.

In [5]:
mapping_col_dict = {'dateCrawled': 'date_crawled',
                    'offerType': 'offer_type',
                    'vehicleType': 'vehicle_type',
                    'yearOfRegistration': 'registration_year',
                    'powerPS': 'power_ps',
                    'monthOfRegistration': 'registration_month',
                    'fuelType': 'fuel_type',
                    'notRepairedDamage': 'unrepaired_damage',
                    'dateCreated': 'ad_created',
                    'nrOfPictures': 'picture_count',
                    'postalCode': 'postal_code',
                    'lastSeen': 'last_viewed'
                   }

autos = autos.rename(mapping_col_dict, axis="columns")

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  picture_count  \
0     peugeot              nein  2016-03-26 00:00:00              0   
1         bmw              nein  2016-04-04 00:00:00              0   
2  volkswagen              nein  2016-03-26 00:00:00              0   
3       smart              nein  2016-03-12 00:00:00              0   
4        ford              nein  2016-04-01 00:00:00              0   

   postal_code          last_viewed  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Taking a closer look at the data that needs cleaning.  
`seller`, `offer_type`, and `unrepaired_damage` only have two unique entries. One of them has 49,999 occurences. We can probably drop these columns altogether.  
`registration_year`, `power_ps`, `postal_code`, and `picture_count` need to be converted to `int`.  
`price` and `odometer` are currently string objects. They need to be stripped of `$` and converted to `int`.
All the date specific columns have a date format. Gonna have to make sure those become the correct `dtype`.  
`registration_month` has values of 1000 and 9999. Those are certianly not years that are possible.

I'm going to start by dropping the useless columns containing 99.99% the same data.

In [6]:
autos.drop(['seller', 'offer_type', 'unrepaired_damage'], axis=1, inplace=True)

autos.describe(include='all')

date_crawled         name  price abtest vehicle_type  \
count                 50000        50000  50000  50000        44905   
unique                48213        38754   2357      2            8   
top     2016-04-02 11:37:04  Ford_Fiesta     $0   test    limousine   
freq                      3           78   1421  25756        12859   
mean                    NaN          NaN    NaN    NaN          NaN   
std                     NaN          NaN    NaN    NaN          NaN   
min                     NaN          NaN    NaN    NaN          NaN   
25%                     NaN          NaN    NaN    NaN          NaN   
50%                     NaN          NaN    NaN    NaN          NaN   
75%                     NaN          NaN    NaN    NaN          NaN   
max                     NaN          NaN    NaN    NaN          NaN   

        registration_year  gearbox      power_ps  model   odometer  \
count        50000.000000    47320  50000.000000  47242      50000   
unique                NaN        2           NaN    245         13   
top                   NaN  manuell           NaN   golf  150,000km   
freq                  NaN    36993           NaN   4024      32424   
mean          2005.073280      NaN    116.355920    NaN        NaN   
std            105.712813      NaN    209.216627    NaN        NaN   
min           1000.000000      NaN      0.000000    NaN        NaN   
25%           1999.000000      NaN     70.000000    NaN        NaN   
50%           2003.000000      NaN    105.000000    NaN        NaN   
75%           2008.000000      NaN    150.000000    NaN        NaN   
max           9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month fuel_type       brand           ad_created  \
count         50000.000000     45518       50000                50000   
unique                 NaN         7          40                   76   
top                    NaN    benzin  volkswagen  2016-04-03 00:00:00   
freq                   NaN     30107       10687                 1946   
mean              5.723360       NaN         NaN                  NaN   
std               3.711984       NaN         NaN                  NaN   
min               0.000000       NaN         NaN                  NaN   
25%               3.000000       NaN         NaN                  NaN   
50%               6.000000       NaN         NaN                  NaN   
75%               9.000000       NaN         NaN                  NaN   
max              12.000000       NaN         NaN                  NaN   

        picture_count   postal_code          last_viewed  
count         50000.0  50000.000000                50000  
unique            NaN           NaN                39481  
top               NaN           NaN  2016-04-07 06:17:27  
freq              NaN           NaN                    8  
mean              0.0  50813.627300                  NaN  
std               0.0  25779.747957                  NaN  
min               0.0   1067.000000                  NaN  
25%               0.0  30451.000000                  NaN  
50%               0.0  49577.000000                  NaN  
75%               0.0  71540.000000                  NaN  
max               0.0  99998.000000                  NaN

**Clean `price` column, convert to `int64`.**  
`dtype` for `value_count` being `int64` refers to the `value_count` as it's OWN object.

In [7]:
print(autos["price"].value_counts())
print(autos["price"].describe())

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$414          1
$79,933       1
$5,198        1
$18,890       1
$16,995       1
Name: price, Length: 2357, dtype: int64
count     50000
unique     2357
top          $0
freq       1421
Name: price, dtype: object


In [8]:
autos["price"] = autos["price"].str.replace('$', '', regex=False).str.replace(',', '', regex=False).astype(int)
autos["price"].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
414         1
79933       1
5198        1
18890       1
16995       1
Name: price, Length: 2357, dtype: int64

**Clean `odometer` column, convert to `int64`.**  
Rename column as `odometer_km`.

In [9]:
print(autos["odometer"].value_counts())
print(autos["odometer"].describe())

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64
count         50000
unique           13
top       150,000km
freq          32424
Name: odometer, dtype: object


In [10]:
autos["odometer"] = autos["odometer"].str.replace('km', '', regex=False).str.replace(',', '', regex=False).astype(int)
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

In [11]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the `odometer_km` and `price` columns.

In [12]:
prices = autos["price"]

# print(prices.unique().shape, '\n')
# print(prices.describe(), '\n')
# print(prices.value_counts().head().sort_index(ascending=False), '\n')
# print(prices.value_counts().head(25).sort_index(ascending=True), '\n')
# print(prices.value_counts().sort_index(ascending=False).head(), '\n')

print(prices.value_counts(bins=[0, 1, 100, 1000, 5000, 10000, 50000, 150000, 99999999]).sort_index(), '\n')


(-0.001, 1.0]              1577
(1.0, 100.0]                319
(100.0, 1000.0]           10103
(1000.0, 5000.0]          20993
(5000.0, 10000.0]          8935
(10000.0, 50000.0]         7873
(50000.0, 150000.0]         166
(150000.0, 99999999.0]       34
Name: price, dtype: int64 



In [13]:
odometers = autos["odometer_km"]

# print(odometers.unique().shape, '\n')
# print(odometers.describe(), '\n')
print(odometers.value_counts().sort_index(ascending=True).head(25), '\n')
print(odometers.value_counts(bins=[0, 4999, 50000, 100000, 150000, 300000]).sort_index(), '\n')



5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64 

(-0.001, 4999.0]            0
(4999.0, 50000.0]        4650
(50000.0, 100000.0]      7756
(100000.0, 150000.0]    37594
(150000.0, 300000.0]        0
Name: odometer_km, dtype: int64 



## I skipped part 5!!!

The `registration_year` column has values ranging `1000` to `9999`. A car can't be registered after the date of the listing (2016). So these are incorrect.  
As for the other direction, it's fairly improbable for a car to have been registered before the first few decades of the 1900s. So we'll take a look at the rows outside `1900-2016` and see if we can just drop these rows.  

In [14]:
reg_year = autos["registration_year"]

pre_1900 = autos[reg_year < 1900]
# pre_1900.describe(include='all')

pre_1900

date_crawled                                              name  \
10556  2016-04-01 06:02:10                                        UNFAL_Auto   
22316  2016-03-29 16:56:41             VW_Kaefer.__Zwei_zum_Preis_von_einem.   
24511  2016-03-17 19:45:11                       Trabant__wartburg__Ostalgie   
32585  2016-04-02 16:56:39                                        UNFAL_Auto   
35238  2016-03-26 13:45:20  Suche_Skoda_Fabia____Skoda_Fabia_Combi_mit_Klima   
49283  2016-03-15 18:38:53                                        Citroen_HY   

       price   abtest vehicle_type  registration_year  gearbox  power_ps  \
10556    450  control          NaN               1800      NaN      1800   
22316   1500  control          NaN               1000  manuell         0   
24511    490  control          NaN               1111      NaN         0   
32585    450  control          NaN               1800      NaN      1800   
35238      0  control          NaN               1500      NaN         0   
49283   7750  control          NaN               1001      NaN         0   

        model  odometer_km  registration_month fuel_type       brand  \
10556     NaN         5000                   2       NaN  mitsubishi   
22316  kaefer         5000                   0    benzin  volkswagen   
24511     NaN         5000                   0       NaN     trabant   
32585     NaN         5000                   2       NaN  mitsubishi   
35238     NaN         5000                   0    benzin       skoda   
49283  andere         5000                   0       NaN     citroen   

                ad_created  picture_count  postal_code          last_viewed  
10556  2016-04-01 00:00:00              0        63322  2016-04-01 09:42:30  
22316  2016-03-29 00:00:00              0        48324  2016-03-31 10:15:28  
24511  2016-03-17 00:00:00              0        16818  2016-04-07 07:17:29  
32585  2016-04-02 00:00:00              0        63322  2016-04-04 14:46:21  
35238  2016-03-26 00:00:00              0        15517  2016-04-04 00:16:54  
49283  2016-03-15 00:00:00              0        66706  2016-04-06 18:47:20

Only 6 rows before 1900. There are `NaN` values for `vehicle_type` and`gearbox`, and odometer readings that make no sense. Also these rows are claiming they were registered before the companies were even founded.

In [19]:
post_2016 = autos[reg_year > 2016]
post_2016.describe(include='all')

date_crawled        name         price abtest vehicle_type  \
count                  1966        1966  1.966000e+03   1966            2   
unique                 1963        1901           NaN      2            2   
top     2016-03-29 20:38:23  Ford_Focus           NaN   test        coupe   
freq                      2           5           NaN   1000            1   
mean                    NaN         NaN  1.608905e+04    NaN          NaN   
std                     NaN         NaN  3.935826e+05    NaN          NaN   
min                     NaN         NaN  0.000000e+00    NaN          NaN   
25%                     NaN         NaN  9.500000e+02    NaN          NaN   
50%                     NaN         NaN  1.950000e+03    NaN          NaN   
75%                     NaN         NaN  4.450000e+03    NaN          NaN   
max                     NaN         NaN  1.234568e+07    NaN          NaN   

        registration_year  gearbox      power_ps model    odometer_km  \
count         1966.000000     1715   1966.000000  1680    1966.000000   
unique                NaN        2           NaN   181            NaN   
top                   NaN  manuell           NaN  golf            NaN   
freq                  NaN     1432           NaN   209            NaN   
mean          2062.464395      NaN     97.425229   NaN  130707.019329   
std            527.272473      NaN    423.857393   NaN   37583.611576   
min           2017.000000      NaN      0.000000   NaN    5000.000000   
25%           2017.000000      NaN      0.000000   NaN  125000.000000   
50%           2017.000000      NaN     75.000000   NaN  150000.000000   
75%           2018.000000      NaN    120.000000   NaN  150000.000000   
max           9999.000000      NaN  16011.000000   NaN  150000.000000   

        registration_month fuel_type       brand           ad_created  \
count          1966.000000      1215        1966                 1966   
unique                 NaN         5          40                   39   
top                    NaN    benzin  volkswagen  2016-03-28 00:00:00   
freq                   NaN       857         498                   80   
mean              4.654120       NaN         NaN                  NaN   
std               3.912061       NaN         NaN                  NaN   
min               0.000000       NaN         NaN                  NaN   
25%               1.000000       NaN         NaN                  NaN   
50%               4.000000       NaN         NaN                  NaN   
75%               8.000000       NaN         NaN                  NaN   
max              12.000000       NaN         NaN                  NaN   

        picture_count   postal_code          last_viewed  
count          1966.0   1966.000000                 1966  
unique            NaN           NaN                 1951  
top               NaN           NaN  2016-04-07 09:46:25  
freq              NaN           NaN                    2  
mean              0.0  47843.092574                  NaN  
std               0.0  25312.288972                  NaN  
min               0.0   1067.000000                  NaN  
25%               0.0  27592.000000                  NaN  
50%               0.0  45967.000000                  NaN  
75%               0.0  66129.250000                  NaN  
max               0.0  99974.000000                  NaN

In [27]:
autos = autos[autos["registration_year"].between(1900, 2016)]

print(autos.describe())

              price  registration_year      power_ps    odometer_km  \
count  4.802800e+04        48028.00000  48028.000000   48028.000000   
mean   9.585252e+03         2002.80351    117.070417  125544.161739   
std    4.843817e+05            7.31085    195.151278   40106.751417   
min    0.000000e+00         1910.00000      0.000000    5000.000000   
25%    1.150000e+03         1999.00000     71.000000  100000.000000   
50%    2.990000e+03         2003.00000    107.000000  150000.000000   
75%    7.400000e+03         2008.00000    150.000000  150000.000000   
max    1.000000e+08         2016.00000  17700.000000  150000.000000   

       registration_month  picture_count   postal_code  
count        48028.000000        48028.0  48028.000000  
mean             5.767760            0.0  50935.867327  
std              3.696802            0.0  25792.079828  
min              0.000000            0.0   1067.000000  
25%              3.000000            0.0  30459.000000  
50%              6

We removed 1972 rows from 50,000 leaving us 48028. Our ranges seem far more plausible now. With the vast majority beteween `1999-2008`.